### DISCUSSION TITLE
How to pass Return key to an application with terminal in raw mode?

### DISCUSSION THREAD

**Discussion by amno1** [microsoft/terminal#17401] [Q&A]
Hi guys; thanks for the software!

I hope you don't mind a question, about input in raw mode.

I have ported a small repl line reader/editor, sort of a very simplified GNU Readline "-likeish" application written in CommonLisp and available only for POSIX systems. It was relatively easy to port it to win32 console API; and everything but one thing works: I can't use raw Enter key to send the line to the reader. I can use Ctrl+Enter or Ctrl+J; actually both work well out of the box, and seem to send CRLF sequence. However, just plain pure Enter seem to never be passed to the application?

When I look in the debug window I can see following sequence of VT codes being printed after hitting Enter:

` ␛[13;28;13;1;32;1_␛[13;28;13;0;32;1_`

I have logged key presses, and I am sure that the application does not see the sole Enter key.  

I have checked [the docs for raw/cooked mode for the console API](https://learn.microsoft.com/en-us/windows/console/high-level-console-modes) and I have seen some of the discussions here, [for example this one](https://github.com/microsoft/terminal/issues/4999) (there is another one similar). It is probably nothing wrong with the Terminal, but just me misunderstanding some flag or something else stupidly simple.

I have done web search and seen some SX pages, and examples around, but I am not really sure what is going on here. I am really not very familiar with low-level input, or rather to say, I am total n00b :). If someone could point me in the right direction it would be really appreciated and helpful. 

Thanks in advance and best regards!

**Comment by lhecker** [comment]
When you press Enter in terminals it sends CR (= `\r` = 0x0d). Ctrl+J and Ctrl+Enter on the other hand send LF (= `\n` = 0x0a) to the application. If Ctrl+Enter works for you but Enter doesn't, I think all you need to do is to swap the `\n` check somewhere in your code with a `\r` check.

By the way, and if you haven't done so already, you can also enable `ENABLE_VIRTUAL_TERMINAL_INPUT` and get a lot closer to how POSIX systems work.

Edit: Just now I realized that if you enable LNM it'll send CRLF on Enter: https://vt100.net/docs/vt510-rm/LNM.html

**Reply by amno1** [reply]
Yes; exactly. The app I am playing with (not written by me) actually already does translate both linefeed and carriage return to its internal symbol. The problem is just that it seems like \r (13) never reaches the repl.

 The thing is a slightly more complicated because I am actually doing this from a Lisp repl (SBCL), so I am not 100% sure if this is caused by something in the console or on the Lisp side. 

[I do have ENABLE_VIRTUAL_TERMINAL_INPUT on and ENABLE_EXTENDED_FLAGS as well](https://github.com/amno1/linedit/blob/linedit-w32/terminal-glue.lisp#L125-L130). I also saw [this discussion here](https://github.com/microsoft/terminal/issues/15167) about adding support for LNM, and I tested options they discussed on the output handle (screen buffer) but nope. I got some interesting effects in some cases :), but Enter never worked. (I have removed symbols I don't use, but they can be seen in some previous commit).

> if you enable LNM it'll send CRLF on Enter

I have tested to print these bytes \e[20h but it didn't do it either. I don't see it documented [among the VT sequences implemented for console](https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences), but terminal perhaps implements it?

I think I'll try with low-level ReadConsoleInput and see if I can read Enter with that one. 

Thank you very much for the answer and for the pointer. The vt100.net is quite useful, I think I'll need it in the future.
